group-coordinator/src/main/java/org/apache/kafka/coordinator/group/generic/GenericGroup.java

```diff
@@ -0,0 +1,1012 @@
+                    allSubscribedTopics.addAll(new HashSet<>(
+                        ConsumerProtocol.deserializeSubscription(buffer, (short) 0).topics()
+                    ));
+                });
+                return Optional.of(allSubscribedTopics);
+            } catch (SchemaException e) {
+                log.warn("Failed to parse Consumer Protocol " + ConsumerProtocol.PROTOCOL_TYPE + ":" +
+                    protocolName.get() + " of group " + groupId + ". " +
+                    "Consumer group coordinator is not aware of the subscribed topics.", e);
+            }
+        }
+
+        return Optional.empty();
+    }
+
+    /**
+     * Update a member.
+     *
+     * @param member              the member.
+     * @param protocols           the list of protocols.
+     * @param rebalanceTimeoutMs  the rebalance timeout in milliseconds.
+     * @param sessionTimeoutMs    the session timeout in milliseconds.
+     * @param future              the future that is invoked once the join phase is complete.
+     */
+    public void updateMember(
+        GenericGroupMember member,
+        List<Protocol> protocols,
+        int rebalanceTimeoutMs,
+        int sessionTimeoutMs,
+        CompletableFuture<JoinGroupResponseData> future
+    ) {
+        decrementSupportedProtocols(member);
+        member.setSupportedProtocols(protocols);
+        incrementSupportedProtocols(member);
+        member.setRebalanceTimeoutMs(rebalanceTimeoutMs);
+        member.setSessionTimeoutMs(sessionTimeoutMs);
+
+        if (future != null && !member.isAwaitingJoin()) {
+            numMembersAwaitingJoinResponse++;
+        } else if (future == null && member.isAwaitingJoin()) {
+            numMembersAwaitingJoinResponse--;
+        }
+        member.setAwaitingJoinFuture(future);
+    }
+
+    /**
+     * Complete the join future.
+     * 
+     * @param member    the member.
+     * @param response  the join response to complete the future with.
+     * @return true if a join future actually completes.
+     */
+    public boolean completeJoinFuture(
+        GenericGroupMember member,
+        JoinGroupResponseData response
+    ) {
+        if (member.isAwaitingJoin()) {
+            member.awaitingJoinFuture().complete(response);
+            member.setAwaitingJoinFuture(null);
+            numMembersAwaitingJoinResponse--;
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * Complete a member's sync future.
+     * 
+     * @param member    the member.
+     * @param response  the sync response to complete the future with.
+     * @return true if a sync future actually completes.
+     */
+    public boolean completeSyncFuture(
+        GenericGroupMember member,
+        SyncGroupResponseData response
+    ) {
+        if (member.isAwaitingSync()) {
+            member.awaitingSyncFuture().complete(response);
+            member.setAwaitingSyncFuture(null);
+            numMembersAwaitingJoinResponse--;
```

Comment:
I think that this is incorrect, isn't it?

Reply:
thanks for the catch! added a check in `testCompleteSyncFuture`